Document the pattern for evolving case classes in a compatible manner (#2662)
scala/docs.scala-lang: _overviews/tutorials/binary-compatibility-for-library-authors.md

in library releases:

* Default parameter values for methods or classes
* Case classes

You can find detailed explanations, runnable examples and tips to maintain binary compatibility in [Binary Compatibility Code Examples & Explanation](https://github.com/jatcwang/binary-compatibility-guide).

Again, we recommend using MiMa to double-check that you have not broken binary compatibility after making changes.

### Changing case class definition in a compatible manner

Sometimes it is desirable to be able to change the definition of a case class (adding and/or removing fields) while still staying compatible with the users of the case class, i.e. not breaking the so called _binary compatibility_.

To achieve that, follow this pattern:
 * make the constructor `private`

>  * define `private` `unapply` function in the companion object (note that by doing that the case class looses the ability to be used in a pattern match)

Reviewer: ```suggestion
 * define a private `unapply` function in the companion object (note that by doing that the case class loses the ability to be used in a pattern match)
```

typo (loses) + "private" can be considered part of "private ... function" so I wouldn't use backticks here (it could also give a false impression that `private unapply` is valid syntax)

Reviewer: (I'd untick the line above as well)

To achieve that, follow this pattern:
 * make the primary constructor private (this makes private the `copy` method of the class)
 * define a private `unapply` function in the companion object (note that by doing that the case class loses the ability to be used as an extractor in match expressions)
 * for all the fields, define `withXXX` methods on the case class that create a new instance with the respective field changed (you can use the private `copy` method to implement them)
 * create a public constructor by defining an `apply` method in the companion object (it can use the private constructor)

Example:

{% tabs case_class_compat_1 class=tabs-scala-version %}

{% tab 'Scala 2' %}

> ~~~ scala

Author: I've noticed that the other tutorials use triple ` instead of triple ~.

Reviewer 2: It does not really matter but ` didn’t work with GitHub’s suggestions.

Author: Fair enough :smile_cat: I don't seem to be able to figure out escaping in GitHub's MarkDown either...

// Mark the primary constructor as private
case class Person private (name: String, age: Int) {

>   // Create withXxx methods for every field, implemented by using the copy method

Author: Scala 2 is a whole different game. I think we also have to make the `copy` method private. And consistently across the changes to the case class.

Reviewer 2: Oh, good point, I forgot that `copy` is not private in Scala 2. I’ve reverted the changes that included Scala 2 then.

// Transform an instance
println(alice.withAge(alice.age + 1)) // Person(Alice, 43)
~~~
{% endtab %}
{% endtabs %}

If you try to use `Person` as an extractor in a match expression, it will fail with a message like “method unapply cannot be accessed as a member of Person.type”. Instead, you can use it as a typed pattern:

{% tabs case_class_compat_1_c class=tabs-scala-version %}
{% tab 'Scala 2' %}
~~~ scala
alice match {
  case person: Person => person.name
}
~~~
{% endtab %}
{% tab 'Scala 3' %}
~~~ scala
alice match
  case person: Person => person.name
~~~
{% endtab %}
{% endtabs %}
Later in time, you can amend the original case class definition to, say, add an optional `address` field. You
 * add a new field `address` and a custom `withAddress` method,
 * add the former constructor signature as a secondary constructor, private to the companion object. This step is necessary because the compilers currently emit the private constructors as public constructors in the bytecode (see [#12711](https://github.com/scala/bug/issues/12711) and [#16651](https://github.com/lampepfl/dotty/issues/16651)).

{% tabs case_class_compat_2 class=tabs-scala-version %}
{% tab 'Scala 2' %}
~~~ scala

> case class Person private (name: String, age: Int, address: Option[String]) {

Author: For Scala 2 we also have to replicate the "old" `apply` function in the companion object.

Reviewer 2: Reverted.

Again, we recommend using MiMa to double-check that you have not broken binary compatibility after making changes.

> ### Changing case class definition in a compatible manner

Reviewer 2: ```suggestion
### Changing a case class definition in a backwards-compatible manner
```

> Sometimes it is desirable to be able to change the definition of a case class (adding and/or removing fields) while still staying compatible with the users of the case class, i.e. not breaking the so called _binary compatibility_.

Reviewer 2: ```suggestion
Sometimes, it is desirable to change the definition of a case class (adding and/or removing fields) while still staying backwards-compatible with the existing usage of the case class, i.e. not breaking the so-called _binary compatibility_.
```

To achieve that, follow this pattern:

>  * make the constructor private

Reviewer 2: ```suggestion
 * make the constructor private (this also makes private the `copy` method of the class)
```

>  * define a private `unapply` function in the companion object (note that by doing that the case class loses the ability to be used in a pattern match)

Reviewer 3: Well usage in patterns is possible e.g. type patterns `case p: Person =>` so it is "only" constructor patterns that are not possible if I'm correct? So `case Person(n,a) => ` would not work as `unapply` is private, but other patterns work. So I think this should be clarified as the reader might not know about all the unapply mechanics under the hood...

Reviewer 3: Edit: I guess extractor patterns could be made to work if the library author provides custom extractors?

Reviewer 3: Anyway, I think it would be good to explain the implications on pattern matching in more detail...

>  * define `withXXX` methods on the case class that create a new instance with the respective field changed

Reviewer 2: ```suggestion
 * for all the fields, define `withXXX` methods on the case class that create a new instance with the respective field changed
```

 * define custom `apply` factory method(s) in the companion object (these can use the private constructor)

Example:

{% tabs case_class_compat_1 %}
{% tab 'Scala 3 Only' %}

```scala
case class Person private (name: String, age: Int):

>   def withName(name: String) = copy(name = name)

Reviewer 2: ```suggestion
  def withName(name: String): Person = copy(name = name)
```

>   def withAge(age: Int) = copy(age = age)

Reviewer 2: ```suggestion
  def withAge(age: Int): Person = copy(age = age)
```

object Person:
  def apply(name: String, age: Int) = new Person(name, age)
  private def unapply(p: Person) = p
```
{% endtab %}
{% endtabs %}

> Later in time, you can ammend the original case class definition. You

Reviewer 2: ```suggestion
Later in time, you can amend the original case class definition. You
```

 * add a new field `address`,
 * add a custom `withAddress` method and
 * add an `apply` factory method to the companion.

{% tabs case_class_compat_2 %}
{% tab 'Scala 3 Only' %}
```scala

> case class Person private (name: String, age: Int, address: String = ""):

Reviewer 2: I would not recommend using an empty `String` to model “the absence of a value”. Instead, I recommend using `Option[String]` and `None`:

```suggestion
case class Person private (name: String, age: Int, address: Option[String]):
```

  ...
  def withAddress(address: String) = copy(address = address)
object Person:
  ...

>   def apply(name: String, age: Int, address: String) = new Person(name, age, address)

Reviewer 2: I think you should also say that the previous `apply` method needs to be adapted:

~~~ diff
- def apply(name: String, age: Int) = new Person(name, age)
+ def apply(name: String, age: Int) = new Person(name, age, None)
~~~

Author: It doesn't, because we manually re-introduce the constructor with the original fields.

> ```

Reviewer 2: IIRC, that change will still trigger a MiMa error because the primary constructor, although private from the point of view of Scala, was not private at the bytecode level (this is an “implementation detail”). Therefore, you have several options here:
- add back the previous constructor signature as a secondary constructor:
  ~~~ scala
  case class Person private (name: String, age: Int, address: Option[String]):
    def this(name: String, age: Int): Person = this(name, age, None)
  ~~~
- add an exclusion filter to the MiMa settings (in such a case, the filter is not 100% safe because there could be existing _Java_ code calling the previous constructor that will break when run with the new class definition)

Author: So it turns out we have to do 3 things now:
 * add a new constructor, private to the companion object, with the original fields
 * add a custom `withAddress` method and
 * add an `apply` factory method to the companion.

This pattern doesn't seem as attractive as it did in the beginning... :confused:

Reviewer 4: Or: Fix MiMa to not flag this pattern. That would be the correct thing to do, which might also help other code.

Reviewer 5: The problem in this case is that Java code would see this as a perfectly normal constructor. No funny name, no synthetic or artifact flag. Just your regular, public constructor.

Author: "Private" primary constructors vs MiMa seems like a topic that needs further discussion.
I've kicked off https://contributors.scala-lang.org/t/private-primary-constructor-vs-mima/6050

{% endtab %}
{% endtabs %}

The original users can use the case class `Person` as before, all the methods that existed before are present unmodified after this change, thus the compatibility with the users is maintained.

> A regular case class not following this pattern would break its users, because by adding a new field some methods (which could be used by somebody else) change, for example `copy` or the constructor itself.

Reviewer 6: "would break its users" sounds a bit strange, don't have an alternative yet though.

Reviewer 3: "would brake usage" or "would break compatible usage" or similar perhaps?

 * define custom `apply` factory method(s) in the companion object (these can use the private constructor)

Example:

{% tabs case_class_compat_1 %}
{% tab 'Scala 3 Only' %}

```scala
case class Person private (name: String, age: Int):
  def withName(name: String): Person = copy(name = name)
  def withAge(age: Int): Person = copy(age = age)
object Person:
  def apply(name: String, age: Int) = new Person(name, age)
  private def unapply(p: Person) = p
```
{% endtab %}
{% endtabs %}

Later in time, you can amend the original case class definition. You
 * add a new field `address`,
 * add a new constructor, private to the companion object, with the original fields
 * add a custom `withAddress` method and
 * add an `apply` factory method to the companion.

{% tabs case_class_compat_2 %}
{% tab 'Scala 3 Only' %}
```scala
case class Person private (name: String, age: Int, address: Option[String]):
  ...
  private[Person] def this(name: String, age: Int): Person = this(name, age, None)
  def withAddress(address: String) = copy(address = address)
object Person:
  ...
  def apply(name: String, age: Int, address: String) = new Person(name, age, Some(address))
```
{% endtab %}
{% endtabs %}

> 

Reviewer 2: ```suggestion

> Note that an alternative solution, instead of adding back the previous constructor signatures as secondary constructors, consists of adding a [MiMa filter](https://github.com/lightbend/mima#filtering-binary-incompatibilities) to simply ignore the problem. Even though the constructors are effectively public in the bytecode, they can’t be called from Scala programs (but they could be called by Java programs). In an sbt build definition you would add the following setting:
> ~~~ scala
> import com.typesafe.tools.mima.core._
> mimaBinaryIssueFilters += ProblemFilters.exclude[DirectMissingMethodProblem]("Person.this")
> ~~~
> Otherwise, MiMa would fail with an error like “method this(java.lang.String,Int)Unit in class Person does not have a correspondent in current version”.
```

// Mark the primary constructor as private
case class Person private (name: String, age: Int):
  // Create withXxx methods for every field, implemented by using the copy method
  def withName(name: String): Person = copy(name = name)
  def withAge(age: Int): Person = copy(age = age)
object Person:
  // Create a public constructor (which uses the primary constructor)
  def apply(name: String, age: Int) = new Person(name, age)
  // Make the extractor private
  private def unapply(p: Person) = p
```
{% endtab %}
{% endtabs %}
This class can be published in a library and used as follows:

~~~ scala
// Create a new instance
val alice = Person("Alice", 42)
// Transform an instance
println(alice.withAge(alice.age + 1)) // Person(Alice, 43)
~~~

If you try to use `Person` as an extractor in a match expression, it will fail with a message like “method unapply cannot be accessed as a member of Person.type”. Instead, you can use it as a typed pattern:

~~~ scala
alice match
  case person: Person => person.name
~~~
Later in time, you can amend the original case class definition to, say, add an optional `address` field. You
 * add a new field `address` and a custom `withAddress` method,

>  * add the former constructor signature as a secondary constructor, private to the companion object. This step is necessary because the public `apply` method in the companion object calls the former constructor, which was effectively public in the bytecode produced by the compiler.

Reviewer 2: ```suggestion
 * add the former constructor signature as a secondary constructor, private to the companion object. This step is necessary because the compilers currently emit the private constructors as public constructors in the bytecode (see [#12711](https://github.com/scala/bug/issues/12711) and [#16651](https://github.com/lampepfl/dotty/issues/16651)).
```

> {% tabs case_class_compat_2 %}
> {% tab 'Scala 3 Only' %}
> ```scala
> case class Person private (name: String, age: Int, address: Option[String]):
>   ...
>   // Add back the former primary constructor signature
>   private[Person] def this(name: String, age: Int) = this(name, age, None)
>   def withAddress(address: Option[String]) = copy(address = address)
> ```
> {% endtab %}
> {% endtabs %}

Reviewer 2: ```suggestion
{% tabs case_class_compat_2 class=tabs-scala-version %}
{% tab 'Scala 2' %}
~~~ scala
case class Person private (name: String, age: Int, address: Option[String]) {
  ...
  // Add back the former primary constructor signature
  private[Person] def this(name: String, age: Int) = this(name, age, None)
  def withAddress(address: Option[String]) = copy(address = address)
}
~~~
{% endtab %}
{% tab 'Scala 3' %}
~~~ scala
case class Person private (name: String, age: Int, address: Option[String]):
  ...
  // Add back the former primary constructor signature
  private[Person] def this(name: String, age: Int) = this(name, age, None)
  def withAddress(address: Option[String]) = copy(address = address)
~~~
{% endtab %}
{% endtabs %}
```

> Note that an alternative solution, instead of adding back the previous constructor signatures as secondary constructors, consists of adding a [MiMa filter](https://github.com/lightbend/mima#filtering-binary-incompatibilities) to simply ignore the problem. Even though the constructors are effectively public in the bytecode, they can’t be called from Scala programs (but they could be called by Java programs). In an sbt build definition you would add the following setting:
> ~~~ scala
> import com.typesafe.tools.mima.core._
> mimaBinaryIssueFilters += ProblemFilters.exclude[DirectMissingMethodProblem]("Person.this")
> ~~~
> Otherwise, MiMa would fail with an error like “method this(java.lang.String,Int)Unit in class Person does not have a correspondent in current version”.
The original users can use the case class `Person` as before, all the methods that existed before are present unmodified after this change, thus the compatibility with the existing usage is maintained.

The new field `address` can be used as follows:

~~~ scala
// The public constructor sets the address to None by default.
// To set the address, we call withAddress:

> val bob = Person("Bob", 21).withAddress(Some("Atlantic ocean"))

Author: @julienrf Thanks for you suggestions. I've applied them all. But I think this part is worth a discussion.

Shouldn't we recommend readers to also include the `apply` methods in the companion object? Because
```scala
Person("Bob", 21).withAddress(Some("Atlantic ocean"))
```
looks a bit clumsy. This looks much better
```scala
Person("Bob", 21, "Atlantic ocean")
```
I know, we save on some boilerplate in the class definition. But we're just pushing the boilerplate onto the users of the class. Is that the right choice to make?

Ideally we'd have our cake and eat it too, but I don't know if it's possible without a new language level feature...

Reviewer 2: I don’t think the pattern I suggested is clumsy. It is similar to the “builder” pattern we often see around.

Anyway, there are probably several possible variations. The one I suggest here provides a public-facing API that (I think) is simpler because it contains fewer overloads of the `apply` methods.

Reviewer 3: Well, documentation space is not scarce :) so I suggest to explain both possibilities: withX or overloaded apply constructors. It's up to the library author. And if the library user is "refused" a more concise constructor then it's just a self-made extension away...
